lib/ex_aws/credentials_ini/file.ex

```diff
@@ -1,17 +1,111 @@
+         ) do
+           case get_sso_role_credentials(sso_start_url, sso_account_id, sso_role_name) do
+             {:ok, sso_creds} -> {:ok, Map.merge(sso_creds, config_credentials)}
+             {:error, _} = err -> err
+           end
+    end
+
+    defp merge_credentials(config_credentials, shared_credentials) do
       Map.merge(config_credentials, shared_credentials)
     end
 
+    defp get_sso_cache_file(sso_start_url) do
+      hash = :crypto.hash(:sha, sso_start_url) |> Base.encode16() |> String.downcase()
+
+      System.user_home()
+      |> Path.join(".aws/sso/cache/#{hash}.json")
+    end
+
+    defp get_sso_role_credentials(sso_start_url, sso_account_id, sso_role_name) do
+      sso_start_url
+      |> get_sso_cache_file()
+      |> File.read()
+      |> parse_sso_cache_file
+      |> check_sso_expiration
+      |> request_sso_role_credentials(sso_account_id, sso_role_name)
+      |> rename_sso_credential_keys
+    end
+
+    defp parse_sso_cache_file({:ok, contents}) do
+      contents
+      |> Jason.decode()
+    end
+
+    defp parse_sso_cache_file(err), do: err
+
+    defp check_sso_expiration({:ok, %{"expiresAt" => expires_at_str} = config}) do
+      with {:ok, expires_at, _} <- DateTime.from_iso8601(expires_at_str)
+      do
+        case DateTime.compare(expires_at, DateTime.utc_now()) do
+          :gt -> {:ok, config}
+          _ -> {:error, "SSO access token is expired, refresh the token with `aws sso login`"}
+        end
+      else
+        {:error, err} -> {:error, "SSO cache file has invalid expiration format: #{err}"}
+      end
+    end
+
+    defp check_sso_expiration(err), do: err
+
+    #TODO: Implement check that verifies if a user has aws_sso and source profile configured
+    # Saw this concern raised in saml2aws using aws go sdk and think it could be helpful
+
+    # TODO: the :ex_aws.request() abstraction that can choose hackney or what users set requires signing
+    # We can't sign yet, so this is written to hackney, but need to revisit alternatives
+    defp request_sso_role_credentials({:ok, %{"accessToken" => access_token, "region" => region}}, account_id, role_name) do
+      with {:ok, %{status_code: 200, headers: _headers, body: body}} <-
+             request(
+               :get,
+               "https://portal.sso.#{region}.amazonaws.com/federation/credentials?account_id=#{account_id}&role_name=#{role_name}",
+               "",
+               [{"x-amz-sso_bearer_token", access_token}]
+             ) do
+
+        body
+        |> Jason.decode()
+      else
+        {:ok, resp} -> {:error, resp}
+        error -> error
+      end
+    end
+
+    defp request_sso_role_credentials(err, _, _), do: err
+
+    defp rename_sso_credential_keys({:ok, %{"roleCredentials" => role_credentials}}) do
+      {:ok, Enum.reduce(role_credentials, %{}, fn
+        {"accessKeyId", v}, acc -> Map.put(acc, :access_key_id, v)
+        {"expiration", v}, acc -> Map.put(acc, :expiration, v)
```

Comment:
not certain that expiration needs to be here

Reply:
iirc the thought was we want it in the final map that gets set because it will actually check if it has been expired and then can renew the credentials. If the SSO token is expired and the credentials are expired that would be a fatal error.